src/errors.rs

```diff
@@ -0,0 +1,100 @@
+use std::path::PathBuf;
+use thiserror::Error;
+
+/// Generic result type for functions in this crate with
+/// a global error class.
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum Error {
+    // General errors
+    #[error("file not found: {path}")]
+    FileNotFound { path: PathBuf },
+    #[error("invalid (non-unicode) characters in path")]
+    NonUnicodePath,
+    #[error("failed to fetch region")]
+    Fetch,
+    #[error("error seeking to file offset")]
+    FileSeek,
+    #[error("error seeking to {contig:?}:{start} in indexed file")]
+    GenomicSeek { contig: String, start: u64 },
+    #[error("sequence {sequence} not found in index")]
+    UnknownSequence { sequence: String },
+    #[error("error setting threads for for file reading")]
+    SetThreads,
+
+    // Errors for faidx
+    #[error("The given position is too large to be converted to i64")]
+    PositionTooLarge,
+
+    // Errors for Tbx
+    #[error("previous iterator generation failed")]
+    NoIter,
+    #[error("truncated tabix record")]
+    TabixTruncatedRecord,
+    #[error("invTabixalid tabix index")]
+    TabixInvalidIndex,
+
+    // Errors for BAM
+    #[error("error parsing CIGAR string: {msg}")]
+    ParseCigar { msg: String },
+    #[error("unexpected CIGAR operation: {msg}")]
+    UnexpectedCigarOperation { msg: String },
+    #[error("error parsing SAM record: {rec}")]
+    ParseSAM { rec: String },
+    #[error("invalid reference path {path}")]
+    InvalidReferencePath { path: PathBuf },
+    #[error("invalid compression level {level}")]
+    InvalidCompressionLevel { level: u32 },
+    #[error("unable to open SAM/BAM/CRAM file at {target}")]
+    BamOpen { target: String },
+    #[error("unable to open SAM/BAM/CRAM index for {target}")]
+    BamInvalidIndex { target: String },
+    #[error("failed to write record (out of disk space?)")]
+    Write,
+    #[error("invalid record in SAM/BAM/CRAM file")]
+    BamInvalidRecord,
+    #[error("truncated record in SAM/BAM/CRAM file")]
+    BamTruncatedRecord,
+    #[error("format of SAM files are not indexable")]
+    NotIndexable,
```

Comment:
I found this wording ambiguous and looked into the htslib code. I would suggest the following wording:
```suggestion
    #[error("format not indexable by htslib (format is detected as something else than SAM/BAM/CRAM)")]
    NotIndexable,
```
The respective code that can produce the return value of `-3` that is this error uses, is here:
https://github.com/samtools/htslib/blob/34ba6c726925efceaa2cc995d7d4b0409907f331/sam.c#L840-L880

Reply:
Solved in ba84f2c2ff670f4be072ebd4b8c467e9b850737b